packages/perspective-viewer-d3fc/src/js/tooltip/tooltip.js

```diff
@@ -0,0 +1,67 @@
+export function tooltip(container, selection, settings) {
+    selection
+        .filter(d => d.baseValue !== d.mainValue)
+        .on("mouseover", function(data) {
+            const tooltipDiv = getTooltipDiv(container);
+            const html = generateHtml(data, settings);
+            showTooltip(container.node(), this, tooltipDiv, html);
+        })
+        .on("mouseout", () => {
+            const tooltipDiv = getTooltipDiv(container);
+            hideTooltip(tooltipDiv);
+        });
+}
+
+function getTooltipDiv(container) {
+    let tooltipDiv = container.select("div.tooltip");
+    if (tooltipDiv.size() === 0) {
+        tooltipDiv = container
+            .append("div")
+            .attr("class", "tooltip")
+            .style("z-index", 3)
+            .style("opacity", 0);
+    }
+    return tooltipDiv;
+}
+
+function generateHtml(data, settings) {
+    const splits = data.key.split("|");
+    let html = [];
```

Comment:
Rather building a html string, you could create a '/html/legend.html' template, and import it into this module. Then you can do something like:

````javascript
// Apply the empty template (only need to do this once when creating the tooltip div)
tooltipDiv.html(tooltipTemplate);

// Then use D3 to update the elements

// Assuming there's a <ul class="cross-values"> in the template
const crossListItems = tooltipDiv.select(".cross-values")
    .selectAll("li")
    .data(data.crossValues);

crossListItems.enter().append("li");
crossListItems.leave().remove();

crossListItems.text(d => d.name);

// etc...
````